File: 79-word-search/79-word-search.cpp
class Solution {
public:
    
    bool dfs(vector<vector<char>>& board,string& word,int r,int c,int wp,int nrow,int ncol,
             vector<vector<int>>& temp){
        
        if(wp==word.size()) return true;
        
        if(r<0 or r>=nrow or c<0 or c>=ncol or word[wp]!=board[r][c] or temp[r][c]==-1){
            return false;
        }
        
        temp[r][c]=-1;
        
        bool ans = dfs(board,word,r+1,c,wp+1,nrow,ncol,temp)
            or dfs(board,word,r-1,c,wp+1,nrow,ncol,temp)
            or dfs(board,word,r,c+1,wp+1,nrow,ncol,temp)
            or dfs(board,word,r,c-1,wp+1,nrow,ncol,temp); 
        
        temp[r][c]=0;
        return ans;
                    
    }
    
    
    
    bool exist(vector<vector<char>>& board, string word) {
        
        int nrow = board.size();
        int ncol = board[0].size();
        
        vector<vector<int>> temp(nrow,vector<int>(ncol,0)); // to mark visited position
        
        int wp = 0; // word pointer;
        
        for(int i=0;i<nrow;i++){
            for(int j=0;j<ncol;j++){
                if(dfs(board,word,i,j,0,nrow,ncol,temp)) return true;
            }
        }
        
        return false;
        
        // time complexity -> (n*m * (4^n)(dfs is calling 4 recursive max n times))
        
        
    }
};

File: 79-word-search/79-word-search.py
class Solution:
    
    def find(self,board,word,wp,i,j,m,n,visited):
        
        if wp==len(word):
            return True
        
        if i<0 or j<0 or i>=m or j>=n or board[i][j]!=word[wp] or visited[i][j]==-1:
            return False
        
        visited[i][j]=-1
        
        ans = (
            self.find(board,word,wp+1,i+1,j,m,n,visited) or
            self.find(board,word,wp+1,i,j+1,m,n,visited) or
            self.find(board,word,wp+1,i,j-1,m,n,visited) or
            self.find(board,word,wp+1,i-1,j,m,n,visited)
        )
        
        visited[i][j]=0
        
        return ans
    
    
    def exist(self, board: List[List[str]], word: str) -> bool:
        m = len(board)
        n = len(board[0])
        
        wp = 0
        
        visited = [[0]*n for _ in range(m)]
        
        for i in range(m):
            for j in range(n):
                
                if board[i][j]==word[0]:
                    
                    if(self.find(board,word,wp,i,j,m,n,visited)):
                        return True
        
        return False
    
                    
                    
    
        